Qemu morello

LawrenceEsswood opened the pull request:
Changes I have made so far for morello. I am going to keep this a feature branch until I clean up most of the TODOs, but it may be worth reviewing as we go, thus the PR.

LawrenceEsswood committed:
Add sail parsing script and its output

LawrenceEsswood committed:
Add arm trickbox

LawrenceEsswood committed:
Move PESBT into cap_register_t. Make ZERO/DDC/PCC more generic.

LawrenceEsswood committed:
Move getpcc helpers into common code

LawrenceEsswood committed:
Move cpu_capreg_state into common code

LawrenceEsswood committed:
Add decompress cap helper

LawrenceEsswood committed:
Add PCC bounds check helper

LawrenceEsswood committed:
Changes to CHERI common code to support AARCH64

LawrenceEsswood committed:
Add PCC WRITABLE/READABLE flags

LawrenceEsswood committed:
Add a couple cap cursor modification helpers

LawrenceEsswood committed:
Fix otype sign related issues for AARCH64

LawrenceEsswood committed:
Make DDC bounds checker into a more general DDC/PCC bounds checker

LawrenceEsswood committed:
Add TCG implementations of some capability operations

LawrenceEsswood committed:
Add basic CHERI support to AARCH64

LawrenceEsswood committed:
Add morello CPU type

LawrenceEsswood committed:
Add morello board hardware

LawrenceEsswood committed:
Add bare-bones support for migrating capability types

jrtc27 commented on hw/arm/Makefile.objs:
```diff
@@ -11,6 +11,7 @@ obj-$(CONFIG_MAINSTONE) += mainstone.o
 obj-$(CONFIG_MICROBIT) += microbit.o
 obj-$(CONFIG_MUSICPAL) += musicpal.o
 obj-$(CONFIG_NETDUINO2) += netduino2.o
+obj-$(CONFIG_MORELLO) += morello.o
```
That's an interesting ordering function

jrtc27 commented on hw/arm/morello.c:
```diff
@@ -0,0 +1,153 @@
+    Error * errp = NULL;
+
+#ifndef TARGET_CHERI
+    error_and_die("Morello board must be used with CHERI");
```
#error and don't compile it in in the first place?

LawrenceEsswood replied:
Done

jrtc27 commented on hw/arm/morello.c:
```diff
@@ -0,0 +1,153 @@
+#endif
+
+    if (strcmp(machine->cpu_type, ARM_CPU_TYPE_NAME("morello")) != 0) {
+        error_and_die("Use the special 'Morello' cpu type for the Morello board.");
```
Why's this necessary? But also the case should match the actual name.

LawrenceEsswood replied:
It will default correctly, but I assume mixing the morello board with anything but the morello CPU is an error. I fixed the case.

jrtc27 commented on default-configs/arm-softmmu.mak:
```diff
@@ -43,3 +43,4 @@ CONFIG_FSL_IMX6UL=y
 CONFIG_SEMIHOSTING=y
 CONFIG_ALLWINNER_H3=y
 CONFIG_ACPI_APEI=y
+CONFIG_MORELLO=y
```
Belongs in aarch64cheri-softmmu.mak (aarch64-softmmu.mak at a push).

LawrenceEsswood replied:
Moved.

jrtc27 commented on softmmu/vl.c:
```diff
@@ -1756,6 +1756,8 @@ static void version(void)
+    printf("Compiled for AARCH64 (with CHERI)\n");
 #else
 #error "INVALID CHERI target"
 #endif
```
The line below says:
```C
    printf("Built with C0 as NULL register\n");
```
which should be  hoisted into the TARGET_MIPS define (it's also _true_ for RISC-V but a bit meaningless), or just deleted given we've relied on this for years now.

arichardson replied:
+1 for delete

LawrenceEsswood replied:
Its gone.

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -274,8 +308,10 @@ typedef struct CPUARMState {
+    AARCH_REG_TYPE elr_el[4]; /* AArch64 exception link regs  */
+    AARCH_REG_TYPE sp_el[5]; /* AArch64 banked stack pointers */
+
+    uint64_t CCTLR_el[4];
```
Only for TARGET_CHERI

LawrenceEsswood replied:
Done

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -1074,11 +1126,18 @@ int fp_exception_el(CPUARMState *env, int cur_el);
 uint32_t sve_zcr_len_for_el(CPUARMState *env, int el);
 
+#ifdef TARGET_CHERI
+    // Morello does not support 32-bit, so might as well optimise everything away
+    #define is_a64(env) 1
```
Having said that, Morello QEMU would be an interesting place to experiment with getting freebsd32 to work with CHERI.

arichardson replied:
I agree, but it would probably require a significant amount of effort. Until then optimizing the checks out seems good to me.

You might also want to remove ` bool aarch64` from CPUARMState for TARGET_CHERI.

LawrenceEsswood replied:
This is one of the things that I am currently unhappy with, doing this was not quite correct. Some places in code still access aarch64 directly (and I think I have gone and fixed all of those as well). Things currently work, but it would probably be preferable to leave things as they were, and just stop the CPU going out of a64 mode (I don't know how arm AARCH64 actually does this, and I have not found anything in the morello spec as to whether trying should be an exception or just just not do anything).

jrtc27 replied:
The base AArch64 spec (in D1.11.2) says this is an illegal return event, which means PSTATE.IL is set to 1, causing an Illegal Execution fault on the next instruction fetch.

(and SPSR_ELx.M[4] is the bit that determines AArch64 (0) vs AArch32 (1), FWIW)

arichardson commented on target/cheri-common/cheri-compressed-cap/.gitrepo:
```diff
@@ -5,8 +5,8 @@
-	parent = c1a4ca2d4527b63fe57ba66d9c9c48f72b5c8e1e
+	branch = morello
+	commit = 6e61e97621caa2132385a357a40b07b1c9ddcda0
+	parent = ca1edb1fcf2dde67d04419e49f5be1b3125ee358
```
This is broken since you rebased, but I can deal with that later.

jrtc27 replied:
I think the best thing to do is likely deal with cheri-compressed-cap separately, bump it normally in dev and then have a Morello PR to then use it.

arichardson commented on configure:
```diff
@@ -8175,12 +8175,19 @@ case "$target_name" in
     gdb_xml_files="arm-core.xml arm-vfp.xml arm-vfp3.xml arm-neon.xml arm-m-profile.xml"
   ;;
-  aarch64|aarch64_be)
+  aarch64|aarch64_be|aarch64cheri)
```
Maybe we should go for morello here?
```suggestion
  aarch64|aarch64_be|morello)
```

LawrenceEsswood replied:
Will change

arichardson commented on default-configs/aarch64cheri-softmmu.mak:
```diff
@@ -0,0 +1,4 @@
+# Default configuration for cheri-softmmu
```
```suggestion
# Default configuration for morello-softmmu
```

arichardson commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -1312,3 +1405,7 @@ void CHERI_HELPER_IMPL(raise_exception_ddc_bounds(CPUArchState *env,
+
+void CHERI_HELPER_IMPL(decompress_cap(CPUArchState *env, uint32_t regndx)) {
+    get_readonly_capreg(env, regndx);
+}
```
```suggestion
}

```

arichardson commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -725,7 +802,8 @@ void CHERI_HELPER_IMPL(cfromptr(CPUArchState *env, uint32_t cd, uint32_t cb,
         cap_register_t result = *cbp;
-        uint64_t new_addr = cbp->cr_base + rt;
+        uint64_t new_addr = rt;
+        if(!(set_addr)) new_addr +=cbp->cr_base;
```
```suggestion
        if (!set_addr)
            new_addr += cbp->cr_base;
```
I generally just run git-clang-format -f before committing. Might not look perfect but at least it's usually consistent.

arichardson commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -703,6 +775,11 @@ void CHERI_HELPER_IMPL(cfromptr(CPUArchState *env, uint32_t cd, uint32_t cb,
     OOB_INFO(cfromptr)->num_uses++;
 #endif
+
+    bool not_null = (cb & CFROMPTR_0_IS_NOT_NULL) != 0;
```
I think adding separate helpers for those variants and having a common implementation function with an additional flags are/multiple booleans would be cleaner than encoding it in the register number.

LawrenceEsswood replied:
This has changed a little since you last looked at it. cheri_jump_and_link is now the factored out version with a flags argument (as you suggest). I am still passing flags through one of the register arguments in CJALR because I didn't want to duplicate it. Should I just duplicate it with an extra flags argument?

arichardson commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -265,6 +316,27 @@ target_ulong CHERI_HELPER_IMPL(cjalr(CPUArchState *env, uint32_t cd,
+        env->pstate ^= PSTATE_C64;
+
+    // Rebuild cached target-specific cheriflags if anything changed
+    if(system_changed | executive_changed | c64_changed)
```
```suggestion
    if (system_changed || executive_changed || c64_changed)
```
logical or for `bool` variables

LawrenceEsswood replied:
Fixed

arichardson commented on qapi/machine.json:
```diff
@@ -27,7 +27,7 @@
 ##
 { 'enum' : 'SysEmuTarget',
-  'data' : [ 'aarch64', 'alpha', 'arm', 'avr', 'cris', 'hppa', 'i386', 'lm32',
+  'data' : [ 'aarch64', 'aarch64cheri', 'alpha', 'arm', 'avr', 'cris', 'hppa', 'i386', 'lm32',
```
I think we should go for 'morello' here in case there is another aarch64cheri in the future.

arichardson commented on scripts/sail_parser.py:
```diff
@@ -0,0 +1,328 @@
+#!/usr/bin/env python3
+
+# TODO: License
```
```suggestion
# SPDX-License-Identifier: BSD-2-Clause
#
# Copyright (c) 2020 Lawrence Esswood
#
# This work was supported by Innovate UK project 105694, "Digital Security by
# Design (DSbD) Technology Platform Prototype".
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
```

arichardson commented on target/arm/arch_dump.c:
```diff
@@ -252,14 +252,14 @@ int arm_cpu_write_elf64_note(WriteCoreDumpFunction f, CPUState *cs,
     }
     note.prstatus.pr_reg.sp = cpu_to_dump64(s, sp);
-    note.prstatus.pr_reg.pc = cpu_to_dump64(s, env->pc);
+    note.prstatus.pr_reg.pc = cpu_to_dump64(s, get_aarch_reg_as_x(&env->pc));
```
```suggestion
    note.prstatus.pr_reg.pc = cpu_to_dump64(s, cpu_get_current_pc_checked(env));
```

arichardson commented on target/arm/arm-semi.c:
```diff
@@ -302,13 +299,7 @@ static target_ulong arm_flen_buf(ARMCPU *cpu)
-    } else {
-        sp = env->regs[13];
-    }
+    target_ulong sp = arm_get_xreg(env, 31);
```
This looks like it breaks 32-bit?

arichardson commented on target/arm/cheri-archspecific.h:
```diff
@@ -0,0 +1,60 @@
+#define ZERO_REG_NUM 32
+
+// LETODO: All of these
+#include "stdio.h"
```
?

arichardson commented on target/arm/cpu.c:
```diff
@@ -40,13 +40,25 @@
 
     if (is_a64(env)) {
-        env->pc = value;
+        set_aarch_reg_value(&env->pc,value);
```
```suggestion
        arm_update_pc(&env->pc, value, /*can_be_unrepresentable=*/false);
```

RISC-V does:
```
static inline void riscv_update_pc(CPURISCVState *env, target_ulong pc_addr,
                                   bool can_be_unrepresentable)
{
#ifdef TARGET_CHERI
    cheri_update_pcc(&env->PCC, pc_addr, can_be_unrepresentable);
#else
    env->pc = pc_addr;
#endif
#ifdef CONFIG_DEBUG_TCG
    env->_pc_is_current = true;
#endif
}
```

arichardson commented on target/arm/cpu.h:
```diff
@@ -227,8 +241,24 @@ typedef struct CPUARMState {
+     * here. */
+    cap_register_t DDC_current;
+    // Holds DDCs 0 through 3 then restricted 0
+    cap_register_t DDCs[5];
```
Maybe like this?
```suggestion
    cap_register_t DDC_EL[4];
    cap_register_t RDDC_EL0;
```

LawrenceEsswood replied:
Keeping it as a single array makes it indexable. The SP bank mirrors this one, so having a single helper the returns an appropriate index for the current mode is neat.

arichardson commented on target/arm/cpu.h:
```diff
@@ -3185,6 +3293,76 @@ typedef CPUARMState CPUArchState;
+// Set a value as if it were an X register (use get_xreg for the GP register file).
+static inline void set_aarch_reg_to_x(AARCH_REG_TYPE* aarch_reg, target_ulong val) {
+#ifdef TARGET_CHERI
+    bzero(aarch_reg, sizeof(AARCH_REG_TYPE));
```
```suggestion
    bzero(aarch_reg, sizeof(*aarch_reg));
```

arichardson reviewed (commented):
Not looked at everything yet, but here are some initial comments.

LawrenceEsswood committed:
Factor out permission checks for load/store better

LawrenceEsswood committed:
Add load/store pair and GCx family of instructions to morello

LawrenceEsswood committed:
Add printf-like instruction logging function

LawrenceEsswood committed:
Fix sail parser missing instructions

Squash me: fix decode script

LawrenceEsswood committed:
Add tcg condition to string function

LawrenceEsswood committed:
Add sync TCG instruction

LawrenceEsswood committed:
Add static tracking of capability decompression

LawrenceEsswood committed:
Give CapRegState a to string helper

LawrenceEsswood committed:
Move around zero reg defines. Create another psuedo-register to use a temporary.

LawrenceEsswood committed:
Make Morello specific changes to cheri utils

LawrenceEsswood committed:
Add another cap debug helper

LawrenceEsswood committed:
Morello specific changes to setting offset and bounds

LawrenceEsswood committed:
Lots more tcg for capability manipulation

LawrenceEsswood committed:
Use the common tcg for setting merged registers to integers

LawrenceEsswood committed:
Pay attention to out of bounds exponants for bounds checks

LawrenceEsswood committed:
build/ccseal/copytype TCG for morello

LawrenceEsswood committed:
Better tracing of changes to registers

LawrenceEsswood committed:
Move some aarch helpers from internal to cpu.h

LawrenceEsswood committed:
Update valid bounds in more places

LawrenceEsswood committed:
Add restricted mode CSRS. PL_ALL redundant.

LawrenceEsswood committed:
General bug fixes for morello instructions

LawrenceEsswood committed:
Fix ifdefs to work for CHERI that is neither riscv nor mips

LawrenceEsswood committed:
Fix sign related issues with atomic operations

LawrenceEsswood committed:
Add helpers to get certain fields without decompression

LawrenceEsswood committed:
Allow atomics on Morello

LawrenceEsswood committed:
Add comments and some missing instructions to arm cheri decode

LawrenceEsswood committed:
Add some bit/byte packing functions

LawrenceEsswood committed:
64-bit memops should never have MO_SIGN

LawrenceEsswood committed:
Too lazy to fix, adding comment

LawrenceEsswood committed:
Add tag set many. Slightly refactor getting / setting multiple tags with atomicity in mind

LawrenceEsswood committed:
Remove some over-zelous asserts for morello

LawrenceEsswood committed:
Fix off by one on bounds check

LawrenceEsswood committed:
Add cap load helper that returns un-squashed tag

LawrenceEsswood committed:
Add some missing gpr_reg_modified calls

LawrenceEsswood committed:
Add CHERI MMU bits to morello

LawrenceEsswood committed:
Add exclusives and multi-tag loads/stores to morello

LawrenceEsswood committed:
TCG for bounds checking normal loads and stores

LawrenceEsswood committed:
Implement alignment and morello exceptions

LawrenceEsswood committed:
Fix some capability checking logic for atomic operations

LawrenceEsswood committed:
Always bounds check with length at least 1

LawrenceEsswood committed:
Move number of lazy regs into target specific header.

LawrenceEsswood committed:
Fix DDC exception number on MIPS

LawrenceEsswood committed:
Remove defunct print

LawrenceEsswood committed:
Assert too strict on AARCH64

LawrenceEsswood committed:
Fix usage of GETPC to only be at top level in helpers

LawrenceEsswood committed:
Define seperate aligned capability type

LawrenceEsswood committed:
Namespace IS_MORELLO

LawrenceEsswood committed:
Add load pair and branch and link instrution to morello.

LawrenceEsswood committed:
Do not require storecap permission if storing untagged values

LawrenceEsswood committed:
SQUASH: migration

arichardson commented on target/cheri-common/cheri_tagmem.c:
```diff
@@ -381,7 +381,10 @@ void *cheri_tagmem_for_addr(RAMBlock *ram, ram_addr_t ram_offset, size_t size)
 
     uint64_t tag = ram_offset >> CAP_TAG_SHFT;
+#ifndef TARGET_AARCH64
+    // AArch64 seems to use different sizes. Might be worth looking into.
```
This sounds like something that could cause issues when booting.

LawrenceEsswood replied:
I have a feeling when I looked at this I also saw a size of zero...

LawrenceEsswood replied:
Possibly at least worth putting some of these TODOS / fix-me's as known issues in the tracker.

arichardson commented on target/mips/cheri-archspecific-earlier.h:
```diff
@@ -0,0 +1,55 @@
+// the zero register in some contexts. We should really have a DDC_REGNUM
+// distinct from ZERO_REG_NUM (which is what CHERI_EXC_REGNUM_DDC currently is)
+// and CHERI_EXC_REGNUM_DDC which is only used for exceptions and nothing else.
+#define CHERI_EXC_REGNUM_DDC 32
```
The previous code used `#define CHERI_EXC_REGNUM_DDC 0 /* TODO: 32 */`
Probably not worth fixing since MIPS will be EOL at some point.

LawrenceEsswood replied:
Yeah, comment above explains my thinking on this. Everything is currently correct, and its only mips with a CHERI_TRUE_EXC_REGNUM_DDC.

uenoB committed:
meson: Set implicit_include_directories to false

Without this, libvixl cannot be compiled with macOS 11.3 SDK due to
include file name conflict (usr/include/c++/v1/version conflicts with
VERSION).

Signed-off-by: Katsuhiro Ueno <uenobk@gmail.com>
Message-Id: <CA+pCdY09+OQfXq3YmRNuQE59ACOq7Py2q4hqOwgq4PnepCXhTA@mail.gmail.com>
Tested-by: Alexander Graf <agraf@csgraf.de>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 75eebe0b1f15464d19a39c4186bfabf328ab601a)

LawrenceEsswood committed:
Add script to annotate instruction traces with source lines

LawrenceEsswood committed:
Improve implementation of sync

LawrenceEsswood committed:
Add userspace logging console command and remove translation cache flushing on mode change

LawrenceEsswood committed:
Add tag memory property to automatically create tags for memory objects

LawrenceEsswood committed:
Add userspace logging to arm

LawrenceEsswood committed:
Enabled CHERI tags and logging on arm virt board

LawrenceEsswood committed:
Add debug assert for lazy capreg static optimisation

LawrenceEsswood committed:
Remove incorrect TCG_CALL_NO_WG flags on helpers that decompress caps

LawrenceEsswood committed:
Add error handling to tag lookup helper

LawrenceEsswood committed:
Add some missing tcg tmp frees

LawrenceEsswood committed:
Fix some cursor sync issues

LawrenceEsswood committed:
Exceptions write to (and decompress) SP

LawrenceEsswood committed:
Cheri-fy arm's TPID registers

jrtc27 committed:
Fix build by deleting seemingly-unnecessary non-existent include

LawrenceEsswood committed:
Fix glaring wrong argument in setmany

LawrenceEsswood committed:
Do CHERI things for DC ZVA

LawrenceEsswood committed:
Fix morello board minimum page size

LawrenceEsswood committed:
Restore state before CHERI exceptions on morello

arichardson committed:
Add a pytest script to run the morello tests

arichardson committed:
Fix out-of-bounds access found by UBSAN

disas_capreg_reset_all() will write past the end of the
cap_compression_states array.

arichardson committed:
Fix TCG temporary leak in gen_cap_get_type_for_copytype

arichardson commented:
#149
